### PR TITLE
Bugfix 3724 -- Make `--modules-folder` more functional

### PR DESCRIPTION
**Pull request by twooster**
**Summary**

The `--modules-folder` flag seems to be a bit broken, seen in #3724 -- bin links were not working. Additionally, the semantics of what it means to have this flag, and where it should be usable, is a bit unclear.

This commit doesn't address the later, but does attempt to address the former. Included:

* bin-links in `install` are now working correctly
* `run` now looks in the `.bin` subfolder if `--modules-folder` is provided
* `check` now checks in `--modules-folder`, if provided

My personal motivation is to be able to move my `node_modules` folder outside of my application for the purposes of having a clean Docker environment when volume-mounting back to host. No more files with odd UIDs, etc. It seems to work anyway if you specify `NODE_PATH`, etc.

**Test plan**

Tests for the binlinks is included. For the rest, I wanted to get feedback on the intent of `--modules-folder`, to ensure I properly understand it, before writing actual tests.

For my purposes, I tested by self-yarning yarn with a modules folder outside of the main path. Two folders, `yarn/`, where my updated code is, and `yarn-test/`, which is another instance. Bootstrapping one from the other and playing around:

```
tony@shtump ~/projects/yarn-test [bugfix-3724]↑⚡
>>> ../yarn/bin/yarn.js --modules-folder alt_modules
yarn install v0.27.0
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
warning fsevents@1.1.1: The platform "linux" is incompatible with this module.
info "fsevents@1.1.1" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/5] Linking dependencies...
warning "eslint-config-fbjs@2.0.0-alpha.1" has unmet peer dependency "eslint-plugin-jasmine@^2.1.0".
warning "eslint-config-fbjs@2.0.0-alpha.1" has unmet peer dependency "eslint-plugin-prefer-object-spread@^1.1.0".
[5/5] Building fresh packages...
Done in 5.94s.
tony@shtump ~/projects/yarn-test [bugfix-3724]↑⚡
>>> ls
CODE_OF_CONDUCT.md  README.md     circle.yml           lib           src
CONTRIBUTING.md     __tests__     end_to_end_tests     lib-legacy    yarn.lock
Dockerfile.dev      alt_modules   flow-typed           package.json
Dockerfile.node7    appveyor.yml  gulpfile.js          resources
LICENSE             bin           jenkins_jobs.groovy  scripts
tony@shtump ~/projects/yarn-test [bugfix-3724]↑⚡
>>> ls alt_modules/.bin/
acorn       eslint        handlebars    loose-envify  sane        sshpk-verify  which
atob        esparse       jest          miller-rabin  semver      strip-bom
babylon     esvalidate    jest-runtime  mkdirp        sha.js      uglifyjs
errno       flow          js-yaml       prettier      shjs        user-home
escodegen   gulp          jsesc         regjsparser   sshpk-conv  uuid
esgenerate  gunzip-maybe  json5         rimraf        sshpk-sign  webpack
tony@shtump ~/projects/yarn-test [bugfix-3724]↑⚡
>>> NODE_PATH=$PWD/alt_modules ../yarn/bin/yarn.js run uuid --modules-folder $PWD/alt_module
s
yarn run v0.27.0
$ "/home/tony/projects/yarn-test/alt_modules/.bin/uuid"
77203659-090e-4d40-9119-28ba8cc8f1f9
Done in 0.25s.
tony@shtump ~/projects/yarn-test [bugfix-3724]↑⚡
>>> ../yarn/bin/yarn.js check      # oops, forgot --modules-folder
yarn check v0.27.0
warning fsevents@1.1.1: The platform "linux" is incompatible with this module.
info "fsevents@1.1.1" is an optional dependency and failed compatibility check. Excluding it from installation.
error "acorn" not installed
error "ansi-styles" not installed
error "assert-plus" not installed
error "async" not installed
error "cliui" not installed
... snip ...
error "homedir-polyfill#parse-passwd" not installed
error "is-unc-path#unc-path-regex" not installed
error Found 776 errors.
info Visit https://yarnpkg.com/en/docs/cli/check for documentation about this command.
tony@shtump ~/projects/yarn-test [bugfix-3724]↑⚡
>>> ../yarn/bin/yarn.js check --modules-folder alt_modules
yarn check v0.27.0
warning fsevents@1.1.1: The platform "linux" is incompatible with this module.
info "fsevents@1.1.1" is an optional dependency and failed compatibility check. Excluding it from installation.
success Folder in sync.
Done in 2.76s.
tony@shtump ~/projects/yarn-test [bugfix-3724]↑⚡
>>>
```

As you can see, the syntax for `run` is still a bit confusing, given that you have to specify the NODE_PATH for it to work, so maybe there's an improvement there. Happy to hear any suggestions and have a discussion.

### CODE REVIEW

File: src/package-linker.js [hunk 1]
```diff
@@ -329,7 +329,7 @@ export default class PackageLinker {
         topLevelDependencies,
         async ([dest, {pkg}]) => {
           if (pkg.bin && Object.keys(pkg.bin).length) {
-            const binLoc = path.join(this.config.cwd, this.config.getFolder(pkg));
+            const binLoc = this.config.modulesFolder || path.join(this.config.cwd, this.config.getFolder(pkg));
```
Comment (bestander) [inline src/package-linker.js]: maybe config.getFolder should have this  condition instead?

**Comment by bestander** [comment]
Thanks for the PR, @twooster!
Lint and a test failed though

**Comment by twooster** [comment]
Yup, I see that. Will get to it hopefully tomorrow. I was having some spurious unrelated failures on my machine, so I punted it up to Circle to see what happened.

**Comment by tailhook** [comment]
@twooster any updates? We have this issue too and it's important for us basically for using yarn in containers too.

File: src/cli/commands/check.js [hunk 2]
```diff
@@ -39,108 +40,90 @@ export async function verifyTreeCheck(
...
+    const searchPath = dep.searchPath;
+    let version: string | boolean = false;
+    let dependencies: ?Dependencies;
+
+    while (searchPath.length) {
+      const manifestLoc = path.join(...searchPath, dep.name);
+
+      if (pkgVersions.has(manifestLoc)) {
+        version = pkgVersions.get(manifestLoc) || missing;
+        if (version !== missing) { // found
```
Comment (BYK) [inline src/cli/commands/check.js]: This part can become:
```
version = plgVersions.get(manifestLoc);
if (!version) {
    break;
}
```

File: src/cli/commands/check.js [hunk 3]
```diff
@@ -39,108 +40,90 @@ export async function verifyTreeCheck(
...
+          pkgVersions.set(manifestLoc, missing);
+        } else if (!await fs.exists(path.join(manifestLoc, 'package.json'))) {
+          pkgVersions.set(manifestLoc, anyVersion);
+        } else {
+          ({version, dependencies} = await config.readManifest(manifestLoc, registryName));
+          pkgVersions.set(manifestLoc, version);
+          break;
+        }
+      }
+      searchPath.pop();
```
Comment (BYK) [inline src/cli/commands/check.js]: Kind of hard to follow this logic. May be can use some comments. Also do we really need all this refactoring? If not let's defer that to a follow-up diff to see exactly what logic changes are needed to make this patch work.

File: src/cli/commands/check.js [hunk 4]
```diff
@@ -39,108 +40,90 @@ export async function verifyTreeCheck(
   const registryName = 'yarn';
   const registry = config.registries[registryName];
   const rootManifest = await config.readManifest(registry.cwd, registryName);
+  const rootModulesFolder = config.modulesFolder || path.join(registry.cwd, registry.folder);
```
Comment (BYK) [inline src/cli/commands/check.js]: I think we should just make `condfig.modulesFolder` to hold the ultimate truth about this.

**Comment by BYK** [comment]
@twooster - need any help from us?

**Comment by BYK** [comment]
@twooster hello again! Don't wanna feel like pressuring you but I think this PR is awesome and there's little way to go to get it merged. Please let us know if you need anything from us or you don't have time to keep working on this so we can set time aside to finish this up :)

**Comment by schmunk42** [comment]
@BYK Would be great if you could finish this, looks like some time has been passed since the last response.

**Comment by BYK** [comment]
>@BYK Would be great if you could finish this, looks like some time has been passed since the last response.

That was my intention but it requires a bit more time than I currently have. If you're interested, I can help. I already reviewed the original code. It needs a merge from master which should be clean now since I did it earlier and Github doesn't report new conflicts.

**Comment by schmunk42** [comment]
> If you're interested, I can help.

I can't really help you here **Facebook, Inc.** I am just a user of your software.

**Comment by thejameskyle** [comment]
Hey @schmunk42, I'm not a Facebook employee, but that's an inappropriate thing to say to people who are. This is an open source project and not a product of Facebook. Facebook isn't even on the license.

We're all in this together, so if you want to improve the software you use, you are encouraged to. Otherwise do not place an additional burden on maintainers.

If you or someone else wants to take this through the finish line, we'd all encourage you to. Otherwise this will have to be closed since contributors are hard at work on many many other things.

Please be kinder to the people who are providing you with free software

**Comment by shouze** [comment]
@schmunk42 @thejameskyle @BYK I've pulled the current PR, I give a try to take into account review feedbacks and make this one merged ASAP 😉 in another PR I will point here soon.

PS: (I'm not a Facebook employee too ^^). That said, not sure that @schmunk42 was an offense, I believe in people kindness & generosity, if effectively makes happen projects like yarn 👼. FOSS is not a peaceful journey BTW.

File: __tests__/commands/_helpers.js [hunk 5]
```diff
@@ -126,7 +127,14 @@ export async function run<T, R>(
   await fs.mkdirp(path.join(cwd, '.yarn-global'));
   await fs.mkdirp(path.join(cwd, '.yarn-link'));
   await fs.mkdirp(path.join(cwd, '.yarn-cache'));
-  await fs.mkdirp(path.join(cwd, 'node_modules'));
+
+  let modulesFolder;
+  if (flags.modulesFolder) {
+    modulesFolder = flags.modulesFolder = path.join(cwd, flags.modulesFolder);
```
[the remark this answers is not in the document]
Comment (shouze) [reply __tests__/commands/_helpers.js]: @BYK I've pulled the actual fork and working on it, would you mean:
```js
const modulesFolder = path.resolve(cwd, flags.modulesFolder || DEFAULT_MODULES_FOLDER);
```

[the remark this answers is not in the document]
Comment (BYK) [reply __tests__/commands/_helpers.js]: That looks exactly like my code example above. Did you change something and I'm missing it? :(

[the remark this answers is not in the document]
Comment (shouze) [reply __tests__/commands/_helpers.js]: Yup, added cwd as 1st arg, subtle change.

[the remark this answers is not in the document]
Comment (BYK) [reply __tests__/commands/_helpers.js]: Ah, of course. Thanks!

**Comment by jnelson** [comment]
Is anyone still working on this? If so, would anyone like assistance getting it done? Looks close and I'm starting to see workarounds (not a good thing).

To those who have started contributing: please acknowledge and let everyone know how it's going. No one wants to duplicate effort, so if you just don't have the time to wrap this up soon it's helpful to be public about it. No shame! Life and work are real. Unblocking contributions is itself a contribution.

@shouze @BYK @twooster

**Comment by delitescere** [comment]
This will be great!

As npm has many ways of altering the location of `node_modules`, it's interesting that `yarn config` is already aware of it:

```sh
/var/www# export NPM_CONFIG_PREFIX=$PWD/.node_modules_deb 
/var/www# yarn config list
yarn config v1.1.0
info yarn config
{ 'version-tag-prefix': 'v',
  'version-git-tag': true,
  'version-git-sign': false,
  'version-git-message': 'v%s',
  'init-version': '1.0.0',
  'init-license': 'MIT',
  'save-prefix': '^',
  'ignore-scripts': false,
  'ignore-optional': false,
  registry: 'https://registry.yarnpkg.com',
  'strict-ssl': true,
  'user-agent': 'yarn/1.1.0 npm/? node/v8.6.0 linux x64',
  lastUpdateCheck: 1507853955885 }
info npm config
{ prefix: '/var/www/.node_modules_deb' }
Done in 0.09s.
```

Looking forward to `yarn install` also being aware of it 💛

**Comment by BYK** [comment]
I'll go ahead and close this PR due to lack of any movement recently. It is already out of date and a fresh start may be easier at this point to get this issue fixed.

If anyone disagrees, feel free to comment here and ping me.

**Comment by simplenotezy** [comment]
This error is annoying :(

**Comment by BenBrostoff** [comment]
Hey @BYK - would you accept PRs that build off this one here? This issue continues to impact a project I'm working on it and it would be nice to get it resolved.
